tower-http/src/set_response_header.rs

```diff
@@ -0,0 +1,271 @@
+            .field("inner", &self.inner)
+            .field("header_name", &self.header_name)
+            .field("override_existing", &self.override_existing)
+            .field("make", &std::any::type_name::<M>())
+            .finish()
+    }
+}
+
+impl<ReqBody, ResBody, S, M> Service<Request<ReqBody>> for SetResponseHeader<S, M>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    M: MakeHeaderValue<ResBody> + Clone,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future, M>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        ResponseFuture {
+            future: self.inner.call(req),
+            header_name: self.header_name.clone(),
+            make: self.make.clone(),
+            override_existing: self.override_existing,
+        }
+    }
+}
+
+/// Response future for [`SetResponseHeader`].
+#[pin_project]
+#[derive(Debug)]
+pub struct ResponseFuture<F, M> {
+    #[pin]
+    future: F,
+    header_name: HeaderName,
+    make: M,
+    override_existing: bool,
+}
+
+impl<F, ResBody, E, M> Future for ResponseFuture<F, M>
+where
+    F: Future<Output = Result<Response<ResBody>, E>>,
+    M: MakeHeaderValue<ResBody>,
+{
+    type Output = F::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.project();
+        let mut res = ready!(this.future.poll(cx)?);
+
+        if let Some(value) = this.make.make_header_value(&res) {
+            if res.headers().contains_key(&*this.header_name) {
+                if *this.override_existing {
+                    res.headers_mut().insert(this.header_name.clone(), value);
+                }
+            } else {
+                res.headers_mut().insert(this.header_name.clone(), value);
+            }
+        }
+
+        Poll::Ready(Ok(res))
+    }
+}
+
+/// Trait for producing header values from responses.
+///
+/// Used by [`SetResponseHeader`].
+///
+/// You shouldn't normally have to implement this trait since its implemented for closures with the
+/// correct type.
+///
+/// It is also implemented directly for `HeaderValue` so if you just want to add a fixed value you
+/// can suply one directly to [`SetResponseHeaderLayer`].
+pub trait MakeHeaderValue<B> {
+    /// Try to create a header value from the response.
+    fn make_header_value(&mut self, response: &Response<B>) -> Option<HeaderValue>;
+}
+
+impl<F, B> MakeHeaderValue<B> for F
+where
+    F: FnMut(&Response<B>) -> Option<HeaderValue>,
+{
+    fn make_header_value(&mut self, response: &Response<B>) -> Option<HeaderValue> {
+        self(response)
+    }
+}
+
+impl<B> MakeHeaderValue<B> for HeaderValue {
+    fn make_header_value(&mut self, _response: &Response<B>) -> Option<HeaderValue> {
+        Some(self.clone())
+    }
+}
```

Comment:
nit/TIOLI: for completeness sake, what do you think about also adding
```suggestion
}

impl<B> MakeHeaderValue<B> for Option<HeaderValue> {
    fn make_header_value(&mut self, _response: &Response<B>) -> Option<HeaderValue> {
        self.clone()
    }
}
```

this might be handy in cases where there's a fixed value that may or may not be added, and whether or not it's added is determined statically. For example, I might have a config file option turns off adding [`Server`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Server) or [`Via`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Via) headers for security reasons. These would be fixed values, but whether or not they're added is determined at runtime.